src/background/service/APIService.js

```diff
@@ -639,13 +642,78 @@ class APIService {
         params.method = TRANSACTION_TYPE.ReclaimEscrow
         return this.submitTxBody(params, tw)
     }
+    /**
+     * withdraw amount when set allowance
+     * @param {*} params 
+     * @returns 
+     */ 
+    setWithdraw = async (params) => {
+        const CONSENSUS_RT_ID = oasis.misc.fromHex(params.toAddress)
+        const consensusWrapper = new oasisRT.consensusAccounts.Wrapper(CONSENSUS_RT_ID);
+        const withdrawWrapper = consensusWrapper.callWithdraw()
+        return this.submitRuntimeBody(params, withdrawWrapper)
+    }
+    /**
+     * set deposit
+     * @param {*} params 
+     */
+    setDeposit= (params)=>{
+        return new Promise( async(resolve,reject)=>{
+            const tw = oasis.staking.allowWrapper()
+            params.method = TRANSACTION_TYPE.StakingAllow
+            params.toAddress = oasis.staking.addressToBech32(await oasis.staking.addressFromRuntimeID(oasis.misc.fromHex(params.runtimeId)))
+            let result = await this.submitTxBody(params, tw,true,()=>this.onBroadcastEnd(params,resolve,reject)).catch(err=>err)
+            if(result&&result.error){
+                reject({error:result.error})
+            }
+        })
+    }
+    onBroadcastEnd= async(params,resolve,reject)=>{
+        const consensusWrapper = new oasisRT.consensusAccounts.Wrapper(oasis.misc.fromHex(params.runtimeId));
+        const depositWrapper = consensusWrapper.callDeposit()
+        return this.submitRuntimeBody(params, depositWrapper,resolve,reject)
+    }
+    submitRuntimeBody= async(params, wrapper,resolve,reject)=>{
+        try {
+            let buildResult = await buildRuntimeTxBody(params,wrapper)
+            let nonce = buildResult.nonce
+            let txWrapper = buildResult.txWrapper
+            let consensusChainContext = buildResult.consensusChainContext
+
+
+            const privateKey = await this.getCurrentPrivateKey()
+            const bytes = hex2uint(privateKey)
+            let signAccount = this.signerFromPrivateKey(bytes)
+            const signer = new oasis.signature.BlindContextSigner(signAccount)
+
+            const signerInfo = ({
+                address_spec: {signature: {ed25519: signer.public()}},
+                nonce: nonce,
+            });
+            txWrapper.setSignerInfo([signerInfo]);
+            await txWrapper.sign([signer], consensusChainContext);
+
+            await submitTx(txWrapper, RETRY_TIME)
+        
+            let u8Hash = await oasis.hash.hash(txWrapper.unverifiedTransaction[0])
+            let hash = oasis.misc.toHex(u8Hash)
+            let config =  getRuntimeConfig(params.runtimeId)
+            if (hash && config.accountType === RUNTIME_ACCOUNT_TYPE.EVM) {
+                this.checkRuntimeTxStatus(hash,params.runtimeId)
+            }
+            resolve && resolve({code:0})
+        } catch (error) {
+            reject && reject({error:error})
```

Comment:
Why not `return {code:0}` and `throw error`, like normal async functions?

Reply:
great , i will fix there

Reply:
@lukaw3d, does the third commit address your point?

Reply:
Yep

Reply:
in the first commit*